File: pyp_means.m
%%% Nonparametric Power-law Data Clustering


function c, clusters = pyp_means(values, counts, lambda, theta)

    %% Initialization
    len = length(values)

    clusters = struct('means', [], 'assignments', [], 'c', 0)

    m1 = dot(values, counts)/len
    clusters.means = [clusters.means; m1] %array of cluster means
    clusters.assignments = [clusters.assignments; 1:len] %array of cluster assignments for each cluster
    
    keyset = 1:len
    valueset = zeros(257,1)
    clusters.map = containers.Map(keyset, valueset) % map for x (index in values) to cluster
    
    %% Loop until convergence
    %%% NEED TO ADD WHILE LOOP
    
    D_r = [];
    for i = 1:len
        x_i = values(i);
        d_ik = [];
        clusters.c = length(clusters.means)
        
        for k = 1:c
            d_ik = [d_ik; (x_i - clusters.means(k))^2];
           
           [min_d, min_idx] = min(d_ik) 
            
           if min_d - theta > lambda - log(c) * theta
               D_r = [D_r; i];
           
           else
               clusters = assign_cluster(i, min_idx, clusters)
     
           end
        end
        
        % re-cluster the unclustered data set D_r
       clusters = re_cluster(D_r, clusters, values, counts, lambda, theta)

       % center agglomeration
       clusters = center_agglomeration(clusters, lambda, theta)

       % update c, means
       clusters.means = calc_cluster_means(clusters, values, counts)
        
    end

    
end




function z = assign_cluster(i, min_idx, clusters)
    c = length(clusters)
   
    len = length(min_idx)
    if len > 1
        k_assigned = min_idx(unidrnd(len))
    end
   
    k_past = clusters.map(i) %find out where it was assigned before
    clusters.map(i) = k_assigned %now update map assignment
    
    % find x_idx in old assignment list and add to new cluster assignment
    
    old_idx = find(clusters.assignments(k_past) == i)
    
    
    
    clusters.assignments(k_assigned) = [clusters.assignments(k_assigned; i]
   
end


function means = calc_cluster_means(cluster, values, counts)

    c = cluster.c
    for k = 1:c
       
        assignments = clusters.assignments(k);
        len = length(assignments)
        sum = 0;
        
        for i = 1:len
            idx = assignments(i)
            sum = sum + values(idx) * counts(udx)
        end
        
        mean = sum/len
        
        clusters.means(k) = mean
    end
end




function clusters = re_cluster(D_r, clusters, values, counts, lambda, theta)

    for i = 1:length(D_r)
       x_idx = D_r(i) 
       x_i = values(x_idx)
       
       d_ik = [];
       d_r = [];
       
       for k = 1:clusters.c
           d_ik = [d_ik; (x - clusters.means(k))^2]
           d_r = [d_r; min(d_ik)]     
       end
       
       d_r = [d_r, D_r]
       [~, sorter] = sort(d_r(:,1), 'descend')
       d_r = d_r(sorter, :)
       
       new_x_idx = d_r(1,2) %x/value index of largest d_k
       
       
       if d_r(1) - theta > lambda - log(c)* theta
           clusters.means = [clusters.means; values(new_x_idx)]
           clusters.c = length(clusters.means)
           assign_cluster(new_x_index, clusters.c, clusters)
           
       else
           [~, min_idx] = min(d_ik)
       
      
       
       
       
       end
        
    end
end

function clusters = center_agglomeration(clusters, lambda, theta)

    % create inter_cluster_d_ij
    % sort (i,j) by d_ij in ascending order
    % check_agglom for each (i,j) pair
    
    c = clusters.c;
    agglom_term = lambda - theta * log(((c+1)^(c+1))/c^c;
    
        
    d = zeros(i,j) %inter-cluster pair-wise squared distance matrix
    
    for i = 1 : c-1
        for j = i+1 : c
            d(i,j) = (clusters.means(i) - clusters.means(j))^2        
        end
    end
    
    
    
    
    
    
    for i = 1:c-1
        
        n1 = length(clusters.assignments(i));

        
        for j = i+1 : c
            n2 = length(clusters.assignments(j));
    
            if d(i,j) < (n1 + n2)/(n1*n2) * agglom_term
                clusters = combine_clusters(clusters, i, j)
                break
            end
    
    
    % if satisfied == 1, combine the two clusters



end



function satisfied = check_agglom(clusters, i, j, lambda, theta)
    
    sq_diff = (clusters.means(i) - clusters.means(j))^2;;
    
    n1 = length(clusters.assignments(i));
    n2 = length(clusters.assignments(j));

    c = clusters.c;
    
    if sq_diff < (n1 + n2)/(n1*n2) * (lambda - theta * log(((c+1)^(c+1))/c^c))
        satisfied = 1;
    
    else
        satisfied = 0;

    end
end


function clusters = combine_clusters(clusters, i, j)
    % combine_clusters will take two clusters and combine them such that
    % all the points in the two clusters are now in a single cluster
    
    % the cluster struct is updated by recomputing the mean, assignments,
    % and c
    
    
    % clusters:  the struct with assignments, means, and c (number of
    % clusters)
    
    % i,j are indices of clusters to combine

    assignments = [clusters.assignments(i), clusters.assignments(j)] %aggregating the point assignments into a single list
    
    sum = 0 %sum for mean re-computation
    
    for a = 1:length(assignments)
        sum = sum + values(a) * counts(a)
    end
    
    mean = sum/a; %newly recomputed mean for single cluster

    % remove the two old clusters
    idx_to_remove = [i,j]
    clusters.assignments(idx_to_remove) = []
    clusters.means(idx_to_remove) = []
    
    % create new, single cluster and update struct
    clusters.means = [clusters.means; mean]
    clusters.assignments = [clusters.assignments; assignments]
    clusters.c = length(clusters.means)
    
    
    % UPDATE clusters.map

end

File: code/pyp_means.py
import pandas as pd
import numpy as np
from clusters import Clusters
from collections import defaultdict
import operator

class Pyp_means(self):

    def __init__(self, freq, lamb, theta):
	self.freq = freq 
	self.lamb = lamb
	self.theta = theta
	self.clusters = pyp_means(self.freq, self.lamb, self.theta)

    def __pyp_means(self, freq, lamb, theta):
	
	## INITIALIZATION
	N = len(values)
	m1 = np.dot(values, counts)/float(N)

	clusters = Clusters()
	clusters.assignments[m1] = []
	clusters.c = 1
	clusters.map = []

	D_r = []
	notConverged = True	
	epsilon = .001
	
	## NEED WHILE LOOP IN CONVERGENCE
	while notConverged:
	    old_clusters = clusters
	# main for loop
	    for x in freq.keys():
		d = dict()

		for cluster_mean in clusters.assignments.keys():
		    d[cluster_mean] = np.power((x - cluster_mean),2) #squared difference between x and cluster mean
		    min_d = min(d.values())
		    new_cluster = min(d, key=d.get)  #get cluster mean with minimum distance from x

		    if min_d - theta > lamb - np.log(clusters.c)*theta:
			D_r.append(x) 

		    else:
			clusters = assign_cluster(x, new_cluster, clusters)

		clusters = re_cluster(D_r, clusters, freq, lamb, theta) #re-cluster the unclustered data set
		clusters = center_agglomeration(clusters, lamb, theta) #center agglomeration
		clusters = update(clusters, freq) #update c, re-compute cluster means
		notConverged = check_convergence(old_clusters, clusters, epsilon)
		
	return clusters

    def __check_convergence(self, old_clusters, new_clusters, epsilon):

	if old_clusters.c != new_clusters.c:
	    return False

	else:
	    diff = np.subtract(new_clusters.assignments.keys(), old_clusters.assignments.keys())
	    for elem in diff:
		if elem > epsilon: 
		    return False
	    return True



    def __assign_clusters(self, x, new_cluster, clusters):

	# find and remove old assignment
	prev_cluster = clusters.map(x)
	clusters.assignments[prev_cluster].remove(x)
	# assign to new cluster
	clusters.map(x) = new_cluster 
	clusters.assignments[new_cluster].append(x)
	return clusters


    def __update(self, clusters, freq):

	# Update c, the number of clusters
	clusters.c = len(clusters.assignments.keys())

	# Update cluster Means
	for cluster_mean, cluster_list in clusters.assignments.iteritems():
	    new_mean = sum([x*freq[x] for x in cluster_list])/len(cluster_list) # calculate new cluster mean
	    clusters.assignments[new_mean] = clusters.assignments.pop(cluster_mean) # update cluster mean

	return clusters


    def __re_cluster(self, D_r, clusters, lamb, theta):

	while D_r:
	    d_r = dict():

	    for x in D_r:
		d = dict()

		for cluster_mean in clusters.assignments.keys():
		    d[cluster_mean] = np.power((x - cluster_mean),2) #squared difference between x and cluster mean
		    min_d = min(d.values())
		    d_r[x] = min_d

	    d_r = sorted(d_r.iteritems(), key=operator.itemgetter(1))	
	    x_1, d_1 = d_r[0]

	    if d_1 - theta > lamb - np.log(clusters.c)*theta:
		# create new cluster where x_1 is mean
		clusters.c += 1
		clusters.assignments[x_1] = [x_1] 
		clusters.map(x_1) = x_1
	
	    else:
		# assign x_1 to existing cluster
		d = dict()
		for cluster_mean in clusters.assignments.keys():
		    d[cluster_mean] = np.power((x - cluster_mean),2) #squared difference between x and cluster mean
		    new_cluster = min(d, key=d.get)			

		clusters = assign_cluster(x_1, new_cluster, clusters)

	    D_r.remove(x_1)
	return clusters

    def __agglomeration_procedure(self, clusters, freq, lamb, theta):
	center_agglom = agglomeration_check(clusters, lamb, theta)
 
	while center_agglom:
	    m1, m2 = center_agglom.pop() 
	    clusters = combine_clusters(clusters, freq, m1, m2)
	    center_agglom = agglomeration_check(clusters, lamb, theta)

	return clusters

    def __agglomeration_check(self, clusters, lamb, theta):
	c = clusters.c	
	agglom_term = lamb - theta*(np.log(np.power(c+1, c+1)/np.power(c, c)))
	means = clusters.assignments.keys()

	d = dict()
	for i in xrange(c-1): 
	    for j = xrange(i+1,c):
		d(i,j) = np.power(means(i) - means(j),2)	

	center_agglom = []
	for m1, m2 in d.keys():
	    n1 = len(clusters.assignments[i])
	    n2 = len(clusters.assignments[j])
	    if d(m1, m2) < (n1+n2)/(n1*n2)*(lamb - theta*(np.log(np.power(c+1, c+1)/np.power(c, c))
		center_agglom.append((m1, m2))

	return center_agglom
    
    def __combine_clusters(clusters, freq, m1, m2):
	cluster1 = clusters.assignments[m1]
	cluster2 = clusters.assignments[m2]
	N = len(cluster1) + len(cluster2)
	# create new cluster
	new_mean = (sum([x*freq[x] for x in cluster1]) + sum([x*freq[x] for x in cluster2]))/float(N) # calculate new cluster mean
	clusters.assignments[new_mean] = cluster1  + cluster2
	# remove 2 old clusters
	clusters.assignments.pop(m1)
	clusters.assignments.pop(m2)

	return clusters	
